BUILD.md

```diff
@@ -0,0 +1,472 @@
+options:
+
+#### 1. Connect to `public-api` locally
+
+```bash
+bin/go-run controller/cmd/public-api
+```
+
+#### 2. Connect to `public-api` in docker-compose
+
+Stop the web service, then run it locally and set the `--api-addr` flag to the
+address of the public API server that's running in your docker environment:
+
+```bash
+docker-compose stop web
+bin/go-run web --api-addr=$DOCKER_IP:8085
+```
+
+#### 3. Connect to `public-api` in Kubernetes
+
+If you are running the public API server in Kubernetes, forward `localhost:8085`
+to the Conduit controller pod:
+
+```bash
+POD_NAME=$(kubectl --namespace=conduit get po --selector=app=controller -o jsonpath='{.items[*].metadata.name}')
+kubectl -n conduit port-forward $POD_NAME 8085:8085
+```
+
+Then connect the local web process to the forwarded port:
+
+```bash
+bin/go-run web --api-addr=localhost:8085
+```
+
+### Webpack dev server
+
+To develop with a webpack dev server, start the server in a separate window:
+
+```bash
+cd web/app
+yarn webpack-dev-server
+```
+
+And then set the `--webpack-dev-server` flag when running the web server:
+
+```bash
+bin/go-run web --webpack-dev-server=http://localhost:8080
+```
+
+To add a JS dependency:
+
+```bash
+cd web/app
+yarn add [dep]
+```
+
+## Proxy
+
+To build and run the Rust proxy:
+
+```bash
+cargo build -p conduit-proxy
+CONDUIT_PROXY_LOG=trace \
+  CONDUIT_PROXY_PUBLIC_LISTENER=tcp://0:5432 \
+  CONDUIT_PROXY_PRIVATE_FORWARD=tcp://127.0.0.1:1234 \
+  CONDUIT_PROXY_CONTROL_URL=tcp://127.0.0.1:8086 \
+  target/debug/conduit-proxy
+```
+
+To connect to a live `proxy-api` at `localhost:8086`:
+
+```bash
+bin/go-run controller/cmd/proxy-api
+```
+
+# Testing
+
+## Go
+
+```bash
+dep ensure && dep prune
+go test ./...
+go vet ./...
+```
+
+## JavaScript
+
+```bash
+cd web/app
+yarn && yarn webpack
+yarn karma start --single-run
+```
+
+## Rust
+
+```bash
+cargo check
+cargo test
+```
+
+# Updating protobuf dependencies
+
+If you make Protobuf changes, run:
+
+```bash
+bin/protoc-go.sh
+```
+
+# Updating Docker dependencies
+
+The Rust proxy and Go Docker images rely on base dependency images with
+hard-coded SHA's:
+
+`gcr.io/runconduit/go-deps` depends on
+- [`Gopkg.lock`](Gopkg.lock)
+- [`Dockerfile-go-deps`](Dockerfile-go-deps)
+
+`gcr.io/runconduit/proxy-deps` depends on
+- [`Cargo.lock`](Cargo.lock)
+- [`proxy/Dockerfile-deps`](proxy/Dockerfile-deps)
+
+If any of these files change, update the Dockerfile SHA's with:
+
+```bash
+GO_DEPS_SHA=$(sh -c ". bin/_tag.sh && go_deps_sha")
+PROXY_DEPS_SHA=$(sh -c ". bin/_tag.sh && proxy_deps_sha")
+
+find . -type f -name 'Dockerfile*' -exec sed -i '' -e 's/gcr\.io\/runconduit\/go-deps:[^ ]*/gcr\.io\/runconduit\/go-deps:'$GO_DEPS_SHA'/g' {} \;
+find . -type f -name 'Dockerfile*' -exec sed -i '' -e 's/gcr\.io\/runconduit\/proxy-deps:[^ ]*/gcr\.io\/runconduit\/proxy-deps:'$PROXY_DEPS_SHA'/g' {} \;
+```
+
+# Go run
+
+Our Go build instructions use a [`bin/go-run`](bin/go-run) script in lieu
```

Comment:
This might make more sense near the top of the file, since it's used in some of the sections above this section. Maybe, in terms of the layout of this file, we could do something like:

* Languages & Frameworks
  * Rust (include info about `cargo`)
  * Go (include info about `dep` and `go-run`)
  * React (include info about `webpack` and `yarn`)
* Components (for each: short description, where it lives in the repo, how to run it from the command line)
  * Data plane
    * proxy
    * proxy-init
    * ...
  * Control plane
    * proxy-api
    * public-api
    * ...
  * Interface layer
    * cli
    * web
* Development environments (local dev covered in previous section)
  * minikube
  * docker-compose
* Testing
  * Rust
  * Go
  * React
* Dependencies
  * Protobuf
  * Docker

Or... something like that. I think they layout you came up with is a good way to represent all of this info too. It's just a lot of info.

Reply:
This is helpful @klingerf . @olix0r had similar guidance around organizing the doc based on programming language use cases, I've taken all this into account with this latest commit.